homeassistant/components/mqtt/discovery.py

```diff
@@ -62,6 +142,29 @@
             return
 
         payload = dict(payload)
+
+        if TOPIC_PREFIX in payload:
+            prefix = payload[TOPIC_PREFIX]
+            for key in payload.keys():
+                payload[key] = payload[key].replace(TOPIC_PREFIX, prefix)
```

Comment:
We shouldn't do the replace like this, as it would also replace something like `hello/~paulus`. Instead, do this:

```python
for key, value in payload.items():
    if key[0] == TOPIC_PREFIX:
        payload[key] = "{}{}".format(TOPIC_PREFIX, value[1:])
```

Reply:
OK, and that limits the topic prefix to the very beginning, maybe that's a reasonable limitation?
Perhaps it also makes sense to only do the replacement if the configuration variable ends with _topic, so something like:
``` python
for key, value in payload.items():
    if value[0] == TOPIC_PREFIX and key.endswith('_topic'):
        payload[key] = "{}{}".format(TOPIC_PREFIX, value[1:])
```

Reply:
Fixed in latest commit.